<html>
<head><script type="text/javascript" src="/static/js/analytics.js?v=1510456756.0" charset="utf-8"></script>

<script type="text/javascript">archive_analytics.values.service='wb';archive_analytics.values.server_name='wwwb-app43.us.archive.org';archive_analytics.values.server_ms=312;</script><script type="text/javascript" src="/static/js/wbhack.js?v=1510456756.0" charset="utf-8"></script>

<script type="text/javascript">
__wbhack.init('https://web.archive.org/web');
</script>
<link rel="stylesheet" type="text/css" href="/static/css/banner-styles.css?v=1510456756.0" />
<link rel="stylesheet" type="text/css" href="/static/css/iconochive.css?v=1510456756.0" />

<!-- End Wayback Rewrite JS Include -->
<title>YAMAP</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<link rel="stylesheet" href="/web/20090828033245cs_/http://www.genomics.ceh.ac.uk/yamap/quickmineoutput.css" type="text/css">
</head>

<!--<BODY bgcolor="#FFFFCC">-->
<body style="background-color:#AA6633;">
<table align="center" style="border:solid;border-width:2px;border-color:navy;background-color:#FFFFCC">
<tr><td>

<table border="2" cellspacing="2" cellpadding="2" bgcolor="#FFFFFF">
  <tr valign="middle">
    <td width="90%"><span class="quickminetitle">YAMAP</span> </td>
    <!-- <td width="31%"><img src="quickmine.png" width="300" height="92"></td> -->
    <td width="10%"><img src="yamap.jpg"></td>
  </tr>
</table>

<p>
<center>
<a href="https://web.archive.org/web/20090828033245/http://www.genomics.ceh.ac.uk/yamap">YAMAP home</a> |
<a href="https://web.archive.org/web/20090828033245/http://envgen.nox.ac.uk/biolinux.html">Bio-Linux</a> |
<a href="https://web.archive.org/web/20090828033245/http://genomics.nox.ac.uk/mm/index.html">Microbial Metagenomics</a>
</center>



<h2><a name="intro">About YAMAP</a></h2>
<p>Yet Another Microbial/Metagenomic Annotation Pipeline/Program is a Perl application created for the <a href="https://web.archive.org/web/20090828033245/http://www.nerc.ac.uk/">NERC</a> <a href="https://web.archive.org/web/20090828033245/http://genomics.nox.ac.uk/mm/index.html">Microbial Metagenomics</a> programme. It is designed to be a user-friendly way for <a href="https://web.archive.org/web/20090828033245/http://envgen.nox.ac.uk/biolinux.html">Bio-Linux</a> users to run a selection of first-pass annotation tools upon metagenomic sequences in order to determine whether these sequences merit more detailed investigation.

<p>See the <a href="https://web.archive.org/web/20090828033245/http://darwin/cgi-bin/viewcvs.cgi/yamap/">CVS</a> for information on changes. The latest version can also be obtained from the apt repository (see <a href="#download">below</a>).


<p>For more information please refer to the sections below:
<ul>
	<li><a href="#download">Get YAMAP</a>
	<li><a href="#manual">YAMAP manual</a>
	<li><a href="#questions">Feedback</a>
</ul>

<hr>

<h2><a name="download">Get YAMAP</a></h2>
<h3>If you have a Debian system</h3>
<p>To install YAMAP you can add the following to your /etc/apt/sources list, then run <b>apt-get install bio-linux-yamap</b>.
<p><pre>deb http://envgen.nox.ac.uk/bio-linux/ unstable bio-linux</pre>
<p>Alternatively, you can download the package directly from <a href="https://web.archive.org/web/20090828033245/http://envgen.nox.ac.uk/bio-linux/dists/unstable/bio-linux/binary-i386/">the repository</a>.

<h3>If you don't have a Debian system</h3>
<p>The only current option is to download the package from <a href="https://web.archive.org/web/20090828033245/http://envgen.nox.ac.uk/bio-linux/dists/unstable/bio-linux/binary-i386/">our repository</a>, and convert it to tar.gz or rpm format with <a href="https://web.archive.org/web/20090828033245/http://www.linuxquestions.org/linux/answers/Jeremys_Magazine_Articles/Alien_Packages_and_Linux_Mirrors">Alien</a>. YAMAP is intended for Bio-Linux users on a specific project, but if there is sufficient demand it may be made available in other formats. If you install from another format you'll need to make sure that all the associated annotation programs are installed. If they aren't, YAMAP will exit and give you a list of what's missing.

<hr>


<h2><a name="manual">YAMAP manual</a></h2>
<ul>
	<li><a href="#input">Input files</a>
	<li><a href="#interface">Annotation configuration</a>
	<li><a href="#output">Output files</a>
	<li><a href="#view">Viewing the output</a>
</ul>
<h3><a name="input">Input files</a></h3>
<p>YAMAP is designed to process FASTA format nucleotide sequence files only, one sequence per file.

<h3><a name="interface">Interface</a></h3>
<h4>Main interface</h4>
<p><table border="0" align="right">
<tr><td>
<p><img src="/web/20090828033245im_/http://www.genomics.ceh.ac.uk/yamap/Screenshot-yamap_v1.02.png" alt="the main window">
</tr></td>
</table>
The interface contains several sections, as listed below:<br>
<ul>
<li><b>File selection buttons:</b> The &ldquo;select sequences&rdquo; button will pop up a file selection dialogue with which you may select sequences to be annotated. You may select multiple sequences by using the shift or control key as you right click upon them. You may delete files from this box by selecting them in the same way and clicking on the &ldquo;delete selected&rdquo; button. 
<li><b>Sequence list box:</b> Contains a list of all the sequence files that are to be annotated. The scrollbar will become active should sufficient names be entered into the box.
<li><b>Annotation program checkboxes:</b> The red checkboxes denote applications that will be run. Next to each is a configuration button which will pop up a separate window allowing you to configure that application.
<li><b>Run and quit buttons:</b> Run will save your selected configuration to ~/.yamap/yamap_run.ini and then run the analysis. Quit will exit the application without saving your configuration.
</ul>

<h4>Application configuration</h4>
<p>All application configuration windows are broadly similar.
<p><table border="0" align="left">
<tr><td>
<p><img src="/web/20090828033245im_/http://www.genomics.ceh.ac.uk/yamap/Screenshot-bigblast.png" alt="the main window">
</tr></td>
</table>
This is a typical application configuration window, in this case for configuring blast (using big-blast.pl) against a database of your choice (in this case, swissprot). Buttons allow the selection of the blast program to use, and other options may be filled in in the boxes. If you are using a Bio-Linux machine then it is recommended that you leave the number of jobs at 2. The "other options" box allows options to be passed directly to blast, and in this case it shows an expect value being specified with the option &ldquo;-e 0.0001&rdquo;

<p>For more details of any of the programs run by YAMAP, please refer to the program's own documentation. This can be accessed by clicking on the documentation button in the configuration window, which will cause a terminal to appear and display the documentation with <b>man</b>, <b>less</b>; or <b>lynx</b>. In general, the <b>space bar</b> and <b>b</b> key will page down or up respectively, and the <b>q</b> key will quit the documentation.

<h3><a name="output">Output files</a></h3>
<p>When YAMAP is run, it will create an output directory in the same directory as the first file in your list of files. It is best practice to make sure all the files you select are in the same directory. The list below shows the files produced when all the appropriate (i.e. not QuickMine) annotation programs were run on a bacterial whole genome in FASTA format, NC_000117 (Chlamydia trachomatis).

<table border="3" align="left">
<tr><td>
<pre>
yamap_out:
NC_000117.embl
NC_000117.qm
NC_000117.out/


yamap_out/NC_000117.out:
NC_000117
NC_000117.bigblast.crunch
NC_000117.bigblast.stdout
NC_000117.bigblast.tab
NC_000117.einverted.stdout
NC_000117.einverted.tab
NC_000117.equicktandem.stdout
NC_000117.etandem.stdout
NC_000117.etandem.tab
NC_000117.glimmer.stdout
NC_000117.glimmer.tab
NC_000117.msatfinder.stdout
NC_000117.msatfinder.tab
NC_000117.palindrome.stdout
NC_000117.palindrome.tab
NC_000117.rbs.output
NC_000117.rbs.tab
NC_000117.selfblast.crunch
NC_000117.selfblast.stdout
NC_000117.selfblast.tab
NC_000117.tan.out
NC_000117.trnascan.out
NC_000117.trnascan.stdout
</pre>
</td></tr>
</table>

<p>The yamap_out directory contains the following:
<table border="0" align="left">
<tr><td>
<ul>
<li>A directory of further output files, NC_000117.out/.
<li>An EMBL formatted sequence file. N.B. this is not appropriate for EMBL submission, as it has not been edited to remove a lot of the &ldquo;junk&rdquo; that may be generated by the annotation programs.
</ul>
<p>The subdirectory yamap_out/NC_000117.out contains:
<ul>
<li>A link to the original file, NC_000117, which is used by Artemis.
<li>A selection of files ending in .tab. These are the feature tables that are viewed by Artemis, and written into the EMBL formatted file one directory up.
<li>Files ending in .crunch. These are the output of MSPcrunch, a blast post-processing program useful when comparing two sequences with <a href="https://web.archive.org/web/20090828033245/http://www.sanger.ac.uk/Software/ACT/">ACT</a>. These are treated like .tab files.
<li>Files ending in .stdout. These are the output the annotation programs produce as they run. If all is successful you can ignore these, but if there is a problem then they should be inspected to find the cause.
<li>Files ending in .out are output files of programs before they have been converted to .tab format &mdash; most users will not need these.
</ul>
</tr></td></table>

<p>No data will be deleted from this directory between runs, so if you wish to run YAMAP again it is a good idea to delete it (with rm -rf), backing it up if necessary (tar czvf yamap_out.tar.gz yamap_out).

<h3><a name="view">Viewing the output</a></h3>
<p>Click on the button that will appear in the program window and Artemis will start. When you close Artemis, it will re-start on the next sequence. If you have run a lot of sequences and get bored with looking at them, you can kill this process off by clicking on the flashing button.
<p>The data that will appear in Artemis is from EMBL formatted file, created from the .tab and .crunch files that have been read in. You can edit the annotations and save any sequence that you particularly like &mdash; please see the <a href="https://web.archive.org/web/20090828033245/http://www.sanger.ac.uk/Software/Artemis/manual/">Artemis manual</a> for more information. 

<h3>Saving the output</h3>
<p>Eventually, it will be necessary to save sequences that you're happy with in /home/db/collector (for Bio-Linux users). More information on this will be added later, once the system is up and running.

<h3>Pfam searching</h3>
<p>Because it is likely that MM users will be annotating many small metagenomic sequences, YAMAP has been designed to use a local Pfam database in order to reduce network traffic. This would normally be kept in /home/db on a <a href="https://web.archive.org/web/20090828033245/http://envgen.nox.ac.uk/biolinux.html">Bio-Linux</a> machine. If you wish to do this, you may download a version of the database <a href="pfam.tar.gz">here</a> and untar in /home/db. Alternative instructions for obtaining and formatting the database are <a href="https://web.archive.org/web/20090828033245/ftp://ftp.sanger.ac.uk/pub/databases/Pfam/Tools/pfam_scan.pl">here</a> and in the pfam_scan documentation that can be accessed from the pfam configuration window.

<h3>Programmes and Parsers</h3>
<p>The table below lists the different programmes used in YAMAP. A link to their webpage and their associated parser, written for YAMAP, is provided.

<p><table border="3" bordercolor="000080" cellpadding="5" cellspacing="2" border="0" width="75%" units="relative" align="center">
<tr><td><b>Executable</b></td><td><b>Description</b></td><td><b>Parser to create files for Artemis</b></td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/quickmine.html">QuickMine</a></td><td><font size="2.5">QuickMine is a suite of Perl scripts capable of the analysis of large volumes of genomic data, using BLAST as the basis.</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/quickmine_parse_artemis.txt">quickmine_parse_artemis.pl</td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://www.sanger.ac.uk/Software/ACT/v4/manual/start.html#COMPARISON-RUN">Big Blast</a></td><td><font size="2.5">A script that blasts sequences against an appropriate local database and generates feature tables and MSP crunch output. The ability to blast against a database of all metagenomic sequences produced by the MM programme will be added. This script was developed at the <a href="https://web.archive.org/web/20090828033245/http://www.sanger.ac.uk/">Sanger Centre</a></td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/ftp://ftp.sanger.ac.uk/pub/pathogens/software/artemis/extra/big_blast.pl">big_blast.pl</a></td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://www.sanger.ac.uk/Software/Pfam/">Pfam</a></td><td><font size="2.5">Protein domain search against a local database</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/parse_pfam.txt">parse_pfam.pl</td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://cbcb.umd.edu/software/glimmer/">Glimmer</a></td><td><font size="2.5">Microbial gene predictor.</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/glimmer_parse_artemis.txt">glimmer_parse_artemis.pl</td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://www.tigr.org/software/genefinding.shtml">RBSfinder (run as part of Glimmer)</a></td><td><font size="2.5">Finds ribosome binding sites for genes in bacterial and archaeal genomes. It is normally run as a post-processor to the Glimmer gene finder</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/rbs_artemis.txt">rbs_parse_artemis.pl</td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://lowelab.ucsc.edu/tRNAscan-SE/">tRNAscan-SE</a></td><td><font size="2.5">tRNA gene predictor.</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/tRNA_parse_artemis.txt">tRNA_parse_artemis.pl</td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://www.bioinf.ceh.ac.uk/msatfinder">Msatfinder</a></td><td><font size="2.5">Looks for microsatellites.</td><td><font size="2.5">Output is in tab format</td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://emboss.sourceforge.net/apps/release/4.0/emboss/apps/einverted.html">Einverted</a></td><td><font size="2.5">Looks for inverted repeats</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/einverted_parse_artemis.txt">einverted_parse_artemis.pl</a></td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://emboss.sourceforge.net/apps/release/4.0/emboss/apps/etandem.html">Etandem</a></td><td><font size="2.5">Looks for tandem repeats.</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/etandem_parse_artemis.txt">etandem_parse_artemis.pl</td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://emboss.sourceforge.net/apps/release/4.0/emboss/apps/palindrome.html">Palindrome</a></td><td><font size="2.5">Looks for inverted repeats.</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/palindrome_parse_artemis.txt">palindrome_parse_artemis.pl</td></tr>
<tr><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://www.genomics.jhu.edu/TransTerm/transterm.html">TransTerm</a></td><td><font size="2.5">Finds rho-independent transcription terminators in bacterial genomes</td><td><font size="2.5"><a href="https://web.archive.org/web/20090828033245/http://genomics.ceh.ac.uk/yamap/parse_transterm.txt">parse_transterm.pl</td></tr>
</table>

<h3><a name="Quirks">Quirks</a></h3>
<ol>
<li>Sometimes a program will run successfully, but will find nothing. It will then pass this information on to the next program in the chain, which will fail due to lack of input. This may sometimes happen with Glimmer, for example, and means that you may find that no CDS regions are found in the output.
<li>The annotation process can be halted whilst it is running by clicking on the flashing button. However, if this is done when big-blast is running then blastall may ignore this command and keep running. If this happens, kill it off by typing &ldquo;killall blastall&rdquo; into a terminal. 
<li>YAMAP is a bit lazy about checking that the user has supplied the correct file, and strange things may happen if the wrong type is used (probably ugly error messages from bioperl).
<li>If you find that some of the configuration options look odd (particularly msatfinder), then delete ~/.yamap.
<li>The output directory will not be deleted if YAMAP is not re-run. If you want to save your results them make a copy of it before running again. However, this does mean that you can keep re-running and over-writing old results or adding new tests.
<li>If you double click on a configuration or a documentation button, then two copies of the configuration or documentation windows will pop up. This is, in fact, a feature rather than a bug, and can be avoided by not double clicking. If you have several configuration windows of the same type open then the configuration saved by YAMAP will be whichever was in the window you last clicked <b>save</b> in.
</ol>


<hr>

<h2><a name="questions">Feedback</a></h2>
<p>If you have any questions or comments, please e-mail <a href="https://web.archive.org/web/20090828033245/mailto:helpdesk@nebc.nox.ac.uk">Tim Booth</a>. I'd be interested to hear of any problems or suggestions for improvements. I don't have much development time but will try to include requested features and will fix any bugs.

<center>
<br><a href="#intro">To the top.</a>
</center>
<hr>


<br>
<center>
<br>
<p><a href="https://web.archive.org/web/20090828033245/http://www.vim.org/"><img src="/web/20090828033245im_/http://www.genomics.ceh.ac.uk/yamap/vim.gif" alt="HTML edited by Vim" border="0"></a>
<br>
<p><a href="https://web.archive.org/web/20090828033245/http://www.gentoo.org/"><img src="/web/20090828033245im_/http://www.genomics.ceh.ac.uk/yamap/powered-by-gentoo3.png" alt="YAMAP wasn't developed on Gentoo Linux, unfortunately, but Gentoo rocks" border="1"></a>
<br>
</center>
</td></tr>
</body>

</html>
